<launch>
    <arg name="package" />
    <arg name="testprefix" />
    <arg name="action_namespace" />
    <arg name="node_configurable" default="false" />
    <arg name="configuration_namespace" default="none" />

    <!-- NodeState::STOPPED -->
    
    <test test-name="$(arg testprefix)_goal_state_aborted_after_unknown_transition" pkg="$(arg package)" type="TransitionsRostest">
        <param name="test_filter" value="goalStateAbortedAfterUnknownTransition" />
        <param name="action_namespace" value="$(arg action_namespace)" />
    </test>

    <test test-name="$(arg testprefix)_node_stopped_init_valid_success" pkg="$(arg package)" type="TransitionsRostest">
        <param name="test_filter" value="nodeStoppedInitValidSuccess" />
        <param name="action_namespace" value="$(arg action_namespace)" />
    </test>

    <!-- NodeState::NODE_CONFIGURED -->
    
    <test test-name="$(arg testprefix)_node_configured_connect_valid_success" pkg="$(arg package)" type="TransitionsRostest">
        <param name="test_filter" value="nodeConfiguredConnectValidSuccess" />
        <param name="action_namespace" value="$(arg action_namespace)" />
    </test>
    
    <test test-name="$(arg testprefix)_node_configured_stop_valid_success" pkg="$(arg package)" type="TransitionsRostest">
        <param name="test_filter" value="nodeConfiguredStopValidSuccess" />
        <param name="action_namespace" value="$(arg action_namespace)" />
    </test>

    <!-- NodeState::COMPONENT_CONNECTED -->
    
    <test test-name="$(arg testprefix)_node_connected_disconnect_valid_success" pkg="$(arg package)" type="TransitionsRostest">
        <param name="test_filter" value="nodeConnectedDisconnectValidSuccess" />
        <param name="action_namespace" value="$(arg action_namespace)" />
    </test>
    
    <test test-name="$(arg testprefix)_node_connected_start_valid_running_success" pkg="$(arg package)" type="TransitionsRostest">
        <param name="test_filter" value="nodeConnectedStartValidRunningSuccess" />
        <param name="action_namespace" value="$(arg action_namespace)" />
    </test>
    
    <!-- NodeState::COMPONENT_DISCONNECTED -->
    
    <test test-name="$(arg testprefix)_node_disconnected_connect_valid_success" pkg="$(arg package)" type="TransitionsRostest">
        <param name="test_filter" value="nodeDisconnectedConnectValidSuccess" />
        <param name="action_namespace" value="$(arg action_namespace)" />
    </test>
    
    <test test-name="$(arg testprefix)_node_disconnected_stop_valid_success" pkg="$(arg package)" type="TransitionsRostest">
        <param name="test_filter" value="nodeDisconnectedStopValidSuccess" />
        <param name="action_namespace" value="$(arg action_namespace)" />
    </test>
    
    <!-- NodeState::COMPONENT_RUNNING -->
    
    <test test-name="$(arg testprefix)_node_running_disconnect_valid_success" pkg="$(arg package)" type="TransitionsRostest">
        <param name="test_filter" value="nodeRunningDisconnectValidSuccess" />
        <param name="action_namespace" value="$(arg action_namespace)" />
    </test>
    
    <test test-name="$(arg testprefix)_node_running_pause_valid_success" pkg="$(arg package)" type="TransitionsRostest">
        <param name="test_filter" value="nodeRunningPauseValidSuccess" />
        <param name="action_namespace" value="$(arg action_namespace)" />
    </test>

    <test test-name="$(arg testprefix)_node_running_stop_and_start_valid_success" pkg="$(arg package)" type="TransitionsRostest">
        <param name="test_filter" value="nodeRunningStopAndStartValidSuccess" />
        <param name="action_namespace" value="$(arg action_namespace)" />
    </test>

    <!-- NodeState::COMPONENT_PAUSED -->
    
    <test test-name="$(arg testprefix)_node_paused_disconnect_valid_success" pkg="$(arg package)" type="TransitionsRostest">
        <param name="test_filter" value="nodePausedDisconnectValidSuccess" />
        <param name="action_namespace" value="$(arg action_namespace)" />
    </test>
    
    <test test-name="$(arg testprefix)_node_paused_resume_valid_success" pkg="$(arg package)" type="TransitionsRostest">
        <param name="test_filter" value="nodePausedResumeValidSuccess" />
        <param name="action_namespace" value="$(arg action_namespace)" />
    </test>

    <group if="$(arg node_configurable)">
        
        <!-- NodeState::STOPPED -->

        <test test-name="$(arg testprefix)_node_stopped_init_valid_fail" pkg="$(arg package)" type="TransitionsRostest">
            <param name="test_filter" value="nodeStoppedInitValidFail" />
            <param name="action_namespace" value="$(arg action_namespace)" />
            <param name="configuration_namespace" value="$(arg configuration_namespace)" />
        </test>        
        
        <!-- NodeState::NODE_CONFIGURED -->

        <test test-name="$(arg testprefix)_node_configured_connect_valid_fail" pkg="$(arg package)" type="TransitionsRostest">
            <param name="test_filter" value="nodeConfiguredConnectValidFail" />
            <param name="action_namespace" value="$(arg action_namespace)" />
            <param name="configuration_namespace" value="$(arg configuration_namespace)" />
        </test>    

        <test test-name="$(arg testprefix)_node_configured_stop_valid_fail" pkg="$(arg package)" type="TransitionsRostest">
            <param name="test_filter" value="nodeConfiguredStopValidFail" />
            <param name="action_namespace" value="$(arg action_namespace)" />
            <param name="configuration_namespace" value="$(arg configuration_namespace)" />
        </test>    
    
        <!-- NodeState::COMPONENT_CONNECTED -->

        <test test-name="$(arg testprefix)_node_connected_disconnect_valid_fail" pkg="$(arg package)" type="TransitionsRostest">
            <param name="test_filter" value="nodeConnectedDisconnectValidFail" />
            <param name="action_namespace" value="$(arg action_namespace)" />
            <param name="configuration_namespace" value="$(arg configuration_namespace)" />
        </test> 

        <test test-name="$(arg testprefix)_node_connected_start_valid_running_fail" pkg="$(arg package)" type="TransitionsRostest">
            <param name="test_filter" value="nodeConnectedStartValidRunningFail" />
            <param name="action_namespace" value="$(arg action_namespace)" />
            <param name="configuration_namespace" value="$(arg configuration_namespace)" />
        </test> 

        <test test-name="$(arg testprefix)_node_connected_start_valid_paused_success" pkg="$(arg package)" type="TransitionsRostest">
            <param name="test_filter" value="nodeConnectedStartValidPausedSuccess" />
            <param name="action_namespace" value="$(arg action_namespace)" />
            <param name="configuration_namespace" value="$(arg configuration_namespace)" />
        </test> 

        <test test-name="$(arg testprefix)_node_connected_start_valid_paused_fail" pkg="$(arg package)" type="TransitionsRostest">
            <param name="test_filter" value="nodeConnectedStartValidPausedFail" />
            <param name="action_namespace" value="$(arg action_namespace)" />
            <param name="configuration_namespace" value="$(arg configuration_namespace)" />
        </test> 

        <!-- NodeState::COMPONENT_DISCONNECTED -->

        <test test-name="$(arg testprefix)_node_disconnected_connect_valid_fail" pkg="$(arg package)" type="TransitionsRostest">
            <param name="test_filter" value="nodeDisconnectedConnectValidFail" />
            <param name="action_namespace" value="$(arg action_namespace)" />
            <param name="configuration_namespace" value="$(arg configuration_namespace)" />
        </test> 

        <test test-name="$(arg testprefix)_node_disconnected_stop_valid_fail" pkg="$(arg package)" type="TransitionsRostest">
            <param name="test_filter" value="nodeDisconnectedStopValidFail" />
            <param name="action_namespace" value="$(arg action_namespace)" />
            <param name="configuration_namespace" value="$(arg configuration_namespace)" />
        </test> 

        <!-- NodeState::COMPONENT_RUNNING -->

        <test test-name="$(arg testprefix)_node_running_disconnect_valid_fail" pkg="$(arg package)" type="TransitionsRostest">
            <param name="test_filter" value="nodeRunningDisconnectValidFail" />
            <param name="action_namespace" value="$(arg action_namespace)" />
            <param name="configuration_namespace" value="$(arg configuration_namespace)" />
        </test> 

        <test test-name="$(arg testprefix)_node_running_pause_valid_fail" pkg="$(arg package)" type="TransitionsRostest">
            <param name="test_filter" value="nodeRunningPauseValidFail" />
            <param name="action_namespace" value="$(arg action_namespace)" />
            <param name="configuration_namespace" value="$(arg configuration_namespace)" />
        </test> 

        <!-- NodeState::COMPONENT_PAUSED -->

        <test test-name="$(arg testprefix)_node_paused_disconnect_valid_fail" pkg="$(arg package)" type="TransitionsRostest">
            <param name="test_filter" value="nodePausedDisconnectValidFail" />
            <param name="action_namespace" value="$(arg action_namespace)" />
            <param name="configuration_namespace" value="$(arg configuration_namespace)" />
        </test> 

        <test test-name="$(arg testprefix)_node_paused_resume_valid_fail" pkg="$(arg package)" type="TransitionsRostest">
            <param name="test_filter" value="nodePausedResumeValidFail" />
            <param name="action_namespace" value="$(arg action_namespace)" />
            <param name="configuration_namespace" value="$(arg configuration_namespace)" />
        </test> 

    </group>

</launch>